actions.php

```diff
@@ -0,0 +1,241 @@
+        'amount' => $amount,
+        'currency' => $currency,
+        'payment_for' => $payment_for,
+        'phone_number' => $phone_number,
+        'variable_amount_id' => $variable_amount_id
+    ));
+
+    // Make the API request to validate the promo code
+    $response = wp_remote_post($api_url, array(
+        'headers' => array(
+            'Content-Type' => 'application/json',
+            'x-api-key' => $api_key // Add the API key header
+        ),
+        'body' => $request_body,
+        'timeout' => 30
+    ));
+
+    // Check if there was an error with the API request
+    if (is_wp_error($response)) {
+        error_log('XPAY Promo Code Error: ' . $response->get_error_message());
+        wp_send_json_error(array('message' => 'Failed to validate promo code'));
+        return;
+    }
+
+    // Retrieve the status code and body of the API response
+    $status_code = wp_remote_retrieve_response_code($response);
+    $body = json_decode(wp_remote_retrieve_body($response), true);
+
+    // Log the response for debugging if debug mode is enabled
+    if ($gateway->get_option('debug') === 'yes') {
+        error_log('XPAY Promo Code Response: ' . print_r([
+            'status' => $status_code,
+            'body' => $body,
+            'headers' => wp_remote_retrieve_headers($response)
+        ], true));
+    }
+
+    // Handle non-200 response status
+    if ($status_code !== 200) {
+        $error_message = isset($body['message']) ? $body['message'] : 'Invalid promo code';
+        wp_send_json_error(array('message' => $error_message));
+        return;
+    }
+
+    // Ensure the response body contains the expected data
+    if (isset($body['data'])) {
+        wp_send_json_success($body['data']);
+    } else {
+        wp_send_json_error(array('message' => 'Invalid response format'));
+    }
+
+    wp_die();
+}
+
+add_action('wp_ajax_store_promocode_id', 'store_promocode_id');
+add_action('wp_ajax_nopriv_store_promocode_id', 'store_promocode_id');
+function store_promocode_id() {
+    check_ajax_referer('validate-promo-code', 'security');
+    
+    // Get the promocode_id and discount_amount from the AJAX request
+    $promocode_id = sanitize_text_field($_POST['promocode_id']);
+    $discount_amount = sanitize_text_field($_POST['discount_amount']);
+    
+    // Store both values in session
+    WC()->session->set('promocode_id', $promocode_id);
+    WC()->session->set('discount_amount', $discount_amount);
+    
+    // Send a success response
+    wp_send_json_success(array('message' => 'Promocode ID and discount amount stored successfully'));
+}
+
+add_action('woocommerce_checkout_order_processed', 'store_xpay_promocode_after_payment', 10, 3);
+function store_xpay_promocode_after_payment($order_id, $posted_data, $order) {
+    $promocode_id = WC()->session->get('promocode_id');
+    $discount_amount = WC()->session->get('discount_amount');
+    error_log("Debug: Storing promocode_id in order meta: $promocode_id");
```

Comment:
[NIPTICK] i prefer not logging any errors/info unless checking for debug option is enabled, you can use sth like jsprint function in the utils file